PlutoHelperAgent/ProjectLockerAndKeychainFunctions.m

```diff
@@ -89,156 +92,123 @@ + (NSArray *) load_data_from_keychain {
     NSURLSession *session = [NSURLSession sessionWithConfiguration:config];
     
-    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:url];
-    request.HTTPMethod = @"GET";
-    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
-    NSDictionary * headers = [NSHTTPCookie requestHeaderFieldsWithCookies:[cookieJar cookies]];
-    [request setAllHTTPHeaderFields:headers];
-    NSLog(@"Request %@", request);
-    NSLog(@"Request %@", request.allHTTPHeaderFields);
+    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:urlComplete];
+    request.HTTPMethod = method;
+    [request setValue:type forHTTPHeaderField:@"Content-Type"];
     
-    NSError *error = nil;
+    if (send_cookie == 1) {
+        
+        NSHTTPCookie *cookie;
+        NSHTTPCookieStorage *cookieJar = [NSHTTPCookieStorage sharedHTTPCookieStorage];
+        for (cookie in [cookieJar cookies]) {
+            NSLog(@"%@", cookie);
+        }
+        
+        NSDictionary * headers = [NSHTTPCookie requestHeaderFieldsWithCookies:[cookieJar cookies]];
+        [request setAllHTTPHeaderFields:headers];
+        
+    }
+    
+    NSLog(@"Request %@", request);
     
-    connectionStatus = 0;
+    communicationStatus = 1;
     
-    if (!error) {
+    if (body != NULL) {
+        
+        NSError *error = nil;
+        NSData *data = [NSJSONSerialization dataWithJSONObject:body
+                                                       options:kNilOptions error:&error];
+        
+        if (!error) {
+            NSURLSessionUploadTask *uploadTask = [session uploadTaskWithRequest:request
+                                                                       fromData:data completionHandler:^(NSData *data,NSURLResponse *response,NSError *error) {
+                                                                           NSString *datastring = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
+                                                                           NSLog(@"Data %@", datastring);
+                                                                           NSLog(@"Response %@", response);
+                                                                           if (error != NULL) {
+                                                                               NSLog(@"Error %@", error);
+                                                                           } else {
+                                                                               communicationStatus = 0;
+                                                                           }
+                                                                           
+                                                                       }];
+            
+            [uploadTask resume];
+        }
+        
+        sleep(1);
```

Comment:
sorry, I am still not comfortable with this.  Even a small server delay (e.g., 5 seconds wait because the server is restarting) could cause the program to become confused, resulting in a very annoyed user.

I think that the issue here is that `NSURLSessionUploadTask` is inherently asynchronous but you're trying to create a synchronous method.

How about adding another parameter to your method, which takes a block similar to `completionHandler` below.  In your `completionHandler` deserialize the json from the response into a Dictionary and then call that block

Reply:
fixed this by converting to asynchronous block